File: 003. Longest Substring Without Repeating Characters/Solution.py
# beats 21.53%
class Solution:
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        maxlength = 0
        length = 0
        mymap = {}
        
        for i in range(len(s)):
            if(s[i] not in mymap):
                mymap[s[i]] = i
                length+=1
            else:
                if maxlength < length:
                    maxlength = length
                l = []
                for key in mymap.keys():
                    if mymap[key] < mymap[s[i]]:
                        l.append(key)
                for key in l:
                    mymap.pop(key)
                length = i- mymap[s[i]]
                mymap[s[i]] = i
            
        if maxlength < length:
            maxlength = length
        
        return maxlength

File: 003. Longest Substring Without Repeating Characters/Solution.cpp
// beats 10%
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int maxlength = 0;
        int length = 0;
        map<char, int> mymap;
        
        for(int i=0; i<s.length(); i++){
            if(mymap.find(s[i]) == mymap.end()){//if s[i] not in map, put s[i] in map, store its index and length goes up.
                mymap[s[i]] = i;
                length++;
            }
            else{//if s[i] in map, updates maxlength, restart counting.
                if(maxlength < length){
                    maxlength = length;
                }
                vector<char> v;
                for(map<char,int>::iterator iter = mymap.begin(); iter!=mymap.end(); iter++){
                    if(iter->second < mymap[s[i]]){
                        v.push_back(iter->first);
                    }
                }
                for(int k=0;k<v.size();k++){
                    mymap.erase(v[k]);
                }
                length = i - mymap[s[i]];
                mymap[s[i]] = i;
            }
        }
        if(maxlength < length){
            maxlength = length;
        }
        
        return maxlength;
    }
};